lodash_exercise/lodash.js

```diff
@@ -1,61 +1,240 @@
 
+		nArr.push(arr[i]);
+	}
+	return nArr;
 }
 
-function union(){
+// Implementation 1
+function takeRight(arr, num=1){
+	var nArr = [];
+	for(var i=0; i<num; i++) {
+		if(i === arr.length)
+			break;
 
+		nArr.push(arr[(arr.length-1) - i]);
+	}
+	return nArr.reverse();
 }
 
-function zipObject(){
+// Implementation 2
+function takeRight(arr, num=1){
+	var nArr = [];
+	while (num > arr.length) {
+		num--;
+	}
 
-}
-
-function includes(){
+	nArr = arr.slice(arr.length - num);
 
+	return nArr;
 }
 
-function sample(){
+function union(...arrays){
+	var rSet = new Set();
+	var nArr = [];
 
-}
+	for(var i=0; i<arrays.length; i++)
+		nArr = nArr.concat(arrays[i]);
 
-function cloneDeep(){
+	for (var j=0; j<nArr.length; j++) {
+		rSet.add(nArr[j]);
+	}
 
+	return Array.from(rSet);
 }
 
-function sumBy(){
+function zipObject(props, values){
+	var nObj = {};
+	for(var i=0; i<props.length; i++) {
+		if (i > values.length -1)
+			nObj[props[i]] = undefined;
 
+		nObj[props[i]] = values[i];
+	}
+	return nObj;
 }
 
-function inRange(){
+function includes(a, val, idx=0){
+	if (typeof a === 'string') {
+		if(a.indexOf(val) !== -1)
+			return true;
+	}
+	else if (Array.isArray(a)) {
+		if (idx > 0) {
+			var nArr = a.slice(idx);
+			if(nArr.indexOf(val) > -1)
+				return true;
+		}
+		else {
+			if(a.indexOf(val) > -1)
+				return true;
+		}
+	}
+	else if (typeof a === 'object') {
+		if (Object.values(a).indexOf(val) > -1)
+			return true;
+	}
+	else if (Array.isArray(a)) {
+		if (idx > 0) {
+			var nArr = a.slice(idx);
+			if(nArr.indexOf(val) > -1)
+				return true;
+		}
+		else
+			if(a.indexOf(val) > -1)
+				return true;
+	}
+
+	return false;
+}
 
+function sample(arr){
+	var num = Math.floor(Math.random() * arr.length); 
+	return arr[num];
 }
 
-function has(){
+function cloneDeep(a){
+	if (Array.isArray(a)) {
+		var nArr = [];
+		for (var i=0; i < a.length; i++) {
+			if (Array.isArray(a[i]) || typeof a[i] === 'object')
+				nArr.push(cloneDeep(a[i]));
+			else
+				nArr.push(a[i]);
+		}
+
+		return nArr;
+	}
+	else if (typeof a === 'object') {
+		var nObj = {};
+		for (key in a) {
+			if ( Array.isArray(a[key]) || typeof a[key] === 'object')
+				nObj[key] = cloneDeep(a[key]);
+		}
+		return nObj;
+	}
+}
 
+function sumBy(arr, second){
+	var sum = 0;
+	if(typeof second === 'function') {
+		for(var i=0; i<arr.length; i++)
+			sum += second(arr[i]);
+	}
+	else if(typeof second === 'string') {
+		for(var i=0; i<arr.length; i++)
+			sum += arr[i][second];
+	}
+
+	return sum;
 }
 
-function omit(){
+function inRange(num, start=0, end){
+	
+	if (start != 0 && end === undefined) {
+		end = start
+		start = 0;
+	}
+
+	if (num < 0)
+		return num > end && num < start;
+	else
+		return num < end && num > start;
+}
 
+function has(obj, gKeys){
+	if(typeof gKeys === 'string')
+		gKeys = gKeys.split('.');
+
+	var presentKeys = [];
+	function helper(a) {
+		for (key in a) {
+			presentKeys.push(key);
+
+			if(typeof a[key] === 'object')
+				helper(a[key]);
+			else
```

Comment:
we shouldn't need an `else return` here since our function will end after these lines